### PR TITLE
Partially implement Indexed samplers in general and specific code in GLSL

### PR DESCRIPTION
**Pull request by FernandoS27**
This PR Implements the tracking algorithm and deduction of indexed samplers. So far they can only be correctly deduced under the next conditions:

- [Non NVN/Nouveau only] Atleast one shader has been executed with atleast 2 normal bound samplers. (This is used to calculate the sizing of bound texture handlers, it uses default values that are compatible with nouveau and NVN, thus only required for other graphic apis).
- The indexed sampler is bound between two normal samplers.

It still does not work properly on the next case:

- The indexed sampler is the last bound sampler in the shader. In this case, we default to a size of only 1 as it's the only guarantee. A method that works on every driver is being tested and will be adressed in a future PR.

### CODE REVIEW

File: src/video_core/shader/node.h [hunk 1]
```diff
@@ -380,6 +443,7 @@ struct MetaTexture {
     Node lod;
     Node component{};
     u32 element{};
+    Node index{};
```
[the remark this answers is not in the document]
Comment (FernandoS27) [reply src/video_core/shader/node.h]: `consistency with the rest of the function`

File: src/video_core/shader/track.cpp [hunk 2]
```diff
@@ -37,6 +38,112 @@ std::pair<Node, s64> FindOperation(const NodeBlock& code, s64 cursor,
 }
 } // Anonymous namespace
 
+std::optional<std::pair<Node, Node>> DecoupleIndirectRead(const OperationNode& operation) {
+    if (operation.GetCode() != OperationCode::UAdd) {
+        return std::nullopt;
+    }
+    Node gpr{};
```
[the remark this answers is not in the document]
Comment (FernandoS27) [reply src/video_core/shader/track.cpp]: code consistency.

File: src/video_core/shader/track.cpp [hunk 3]
```diff
@@ -37,6 +38,112 @@ std::pair<Node, s64> FindOperation(const NodeBlock& code, s64 cursor,
...
+    Node offset{};
```
[the remark this answers is not in the document]
Comment (FernandoS27) [reply src/video_core/shader/track.cpp]: code consistency.

File: src/video_core/shader/shader_ir.h [hunk 4]
```diff
@@ -321,7 +328,7 @@ class ShaderIR final {
                               std::optional<SamplerInfo> sampler_info = std::nullopt);
 
     /// Accesses a texture sampler for a bindless texture.
-    const Sampler* GetBindlessSampler(Tegra::Shader::Register reg,
+    const Sampler* GetBindlessSampler(Tegra::Shader::Register reg, Node& index_var,
```
[the remark this answers is not in the document]
Comment (FernandoS27) [reply src/video_core/shader/shader_ir.h]: Breaks consistency with Bound Samplers and makes it more painful to decide for one or another.

File: src/video_core/shader/shader_ir.h [hunk 5]
```diff
@@ -191,6 +195,7 @@ class ShaderIR final {
     };
 
     void Decode();
+    void PostDecode();
```
[the remark this answers is not in the document]
Comment (FernandoS27) [reply src/video_core/shader/shader_ir.h]: I prefer to keep it general. I don't feel we should have a Deduce function for each thing we deduce, just a full deduction step.

**Comment by bunnei** [comment]
EA conflicts

File: src/video_core/shader/decode.cpp [hunk 6]
```diff
@@ -33,6 +34,52 @@ constexpr bool IsSchedInstruction(u32 offset, u32 main_offset) {
     return (absolute_offset % SchedPeriod) == 0;
 }
 
+void DeduceTextureHandlerSize(VideoCore::GuestDriverProfile* gpu_driver,
+                              std::list<Sampler>& used_samplers) {
```
Comment (lioncash) [inline src/video_core/shader/decode.cpp]:
```suggestion
                              const std::list<Sampler>& used_samplers) {
```

File: src/video_core/shader/decode.cpp [hunk 7]
```diff
@@ -33,6 +34,52 @@ constexpr bool IsSchedInstruction(u32 offset, u32 main_offset) {
...
+        bound_offsets.emplace_back(sampler.GetOffset());
+    }
+    if (count > 1) {
+        gpu_driver->DeduceTextureHandlerSize(std::move(bound_offsets));
+    }
+}
+
+std::optional<u32> TryDeduceSamplerSize(Sampler& sampler_to_deduce,
+                                        VideoCore::GuestDriverProfile* gpu_driver,
+                                        std::list<Sampler>& used_samplers) {
```
Comment (lioncash) [inline src/video_core/shader/decode.cpp]:
```suggestion
                                        const std::list<Sampler>& used_samplers) {
```

File: src/video_core/shader/decode.cpp [hunk 8]
```diff
@@ -33,6 +34,52 @@ constexpr bool IsSchedInstruction(u32 offset, u32 main_offset) {
...
+        }
+        ++count;
+        bound_offsets.emplace_back(sampler.GetOffset());
+    }
+    if (count > 1) {
+        gpu_driver->DeduceTextureHandlerSize(std::move(bound_offsets));
+    }
+}
+
+std::optional<u32> TryDeduceSamplerSize(Sampler& sampler_to_deduce,
```
Comment (lioncash) [inline src/video_core/shader/decode.cpp]:
```suggestion
std::optional<u32> TryDeduceSamplerSize(const Sampler& sampler_to_deduce,
```

File: src/video_core/shader/node.h [hunk 9]
```diff
@@ -276,16 +283,72 @@ class Sampler {
...
 
+    constexpr bool IsIndexed() const {
+        return is_indexed;
+    }
+
+    constexpr u32 Size() const {
+        return size;
+    }
+
+    void SetSize(u32 new_size) {
```
Comment (lioncash) [inline src/video_core/shader/node.h]: This can also be `constexpr`

File: src/video_core/shader/node.h [hunk 10]
```diff
@@ -488,6 +552,19 @@ class GprNode final {
     Tegra::Shader::Register index{};
 };
 
+/// A custom variable
+class CustomVarNode final {
+public:
+    explicit constexpr CustomVarNode(u32 index) : index{index} {}
+
+    u32 GetIndex() const {
```
Comment (lioncash) [inline src/video_core/shader/node.h]: This can also be constexpr

**Pull request merged**